HW2/P2/parallel_vector.pyx

```diff
@@ -71,20 +79,41 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] dest,
                              int repeat):
    cdef:
-       int idx, r
+       int idx, r, ind1, ind2
        omp_lock_t *locks = get_N_locks(counts.shape[0])
 
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
+   # &locks is an array of 1000 omp locks, with each lock corresponding to each position in the count array
+   # note: counts begins as a numpy array with index = value
+   # note: src and dest represent arrays of random indices of the counts array
+   
    with nogil:
+       # execute the following script 100 times (repeat = 100)
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           # loop through each index in src (and dest; total = 1,000,000) - use four threads here
+           for idx in prange(src.shape[0], num_threads=4):
```

Comment:
Interesting, you used the prange on the inner loop and I used it on the outer loop. I didn't consider using it on the inner loop as I thought you would want to run parallelly across the repeats. Maybe it makes more sense to implement this way, I wonder which implementation has better performance.